test/cases.py

```diff
@@ -208,3 +223,70 @@ def data():
         NON_DETERMINISTIC_CASES.append(case_with_device)
 
 ADJOINT_CASES = [False, True]
+
+
+DICT_CASES_NO_DEVICE = [
+    # Cross entropy
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=2),
+        "loss_func": lambda: CrossEntropyLoss(reduction="mean"),
+        "data": lambda: [
+            (UserDict({"x": rand(3, 10)}), classification_targets((3,), 2)),
+            (UserDict({"x": rand(4, 10)}), classification_targets((4,), 2)),
+        ],
+        "seed": 0,
+    },
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=2),
+        "loss_func": lambda: CrossEntropyLoss(reduction="mean"),
+        "data": lambda: [
+            ({"x": rand(3, 10)}, classification_targets((3,), 2)),
+            ({"x": rand(4, 10)}, classification_targets((4,), 2)),
+        ],
+        "seed": 0,
+    },
+    # BCE
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=1),
+        "loss_func": lambda: BCEWithLogitsLoss(reduction="mean"),
+        "data": lambda: [
+            (UserDict({"x": rand(3, 10)}), binary_classification_targets((3, 1))),
+            (UserDict({"x": rand(4, 10)}), binary_classification_targets((4, 1))),
+        ],
+        "seed": 0,
+    },
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=1),
+        "loss_func": lambda: BCEWithLogitsLoss(reduction="mean"),
+        "data": lambda: [
+            ({"x": rand(3, 10)}, binary_classification_targets((3, 1))),
+            ({"x": rand(4, 10)}, binary_classification_targets((4, 1))),
+        ],
+        "seed": 0,
+    },
+    # MSE
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=2),
+        "loss_func": lambda: MSELoss(reduction="mean"),
+        "data": lambda: [
+            (UserDict({"x": rand(3, 10)}), regression_targets((3, 2))),
+            (UserDict({"x": rand(4, 10)}), regression_targets((4, 2))),
+        ],
+        "seed": 0,
+    },
+    {
+        "model_func": lambda: ModelWithDictInput(num_classes=2),
+        "loss_func": lambda: MSELoss(reduction="mean"),
+        "data": lambda: [
+            ({"x": rand(3, 10)}, regression_targets((3, 2))),
+            ({"x": rand(4, 10)}, regression_targets((4, 2))),
+        ],
+        "seed": 0,
+    },
+]
```

Comment:
Could just append to `CASES_NO_DEVICE`, then we can proceed without having to duplicate the test functions.

Reply:
There are a couple of reasons:
1. We don't need to retest everything --- the only thing that needs to be tested is to validate that dict-like inputs are accepted. I made no other changes beyond the specification of the model's inputs.
2. One must change `examples/functorch.py` to reuse the old test cases.

@f-dangel @runame let me know what you think. I'm fine with either way but leaning towards separating the `DICT_CASES` since the purpose is just to validate the input format, and retest every other cases seem excessive.

Reply:
Re 1: I would propose to significantly reduce the number of test cases when using dict-based data loaders.
Re 2: Since we are adding a new flag `batch_size_fn` to the base class of all linear operators, it will make sense to introduce this argument into all functions related to testing/examples.